flang/lib/Lower/ConvertType.cpp

```diff
@@ -139,10 +140,11 @@ struct TypeBuilder {
       TODO("Assumed rank expression type lowering");
 
     // LOGICAL, INTEGER, REAL, COMPLEX, CHARACTER
-    auto baseType = genFIRType(context, category, dynamicType->kind());
+    auto baseType = category == Fortran::common::TypeCategory::Character
+                        ? genCharacterType(context, dynamicType->kind(),
+                                           getCharacterLength(expr))
+                        : genFIRType(context, category, dynamicType->kind());
```

Comment:
Can we get rid of the ternary? Something more like
```c
= genFIRType(context, category, dynamicType->kind(), getLenParams(category, expr));
```
where the last would return an empty `ArrayRef` if there are no LEN type parameters.

Reply:
Thanks, that is nicer and that will also help with derived types (though I think the RecordType is not intended to represent compile time constant length parameters, so having an `ArrayRef` vs an llvm::Optional might be overkill, but Iet's wait derived type lowering to decide how best to polish this even more).

Reply:
Each derived type is uniquely defined per its distinct KIND parameters.

My example, though, was in relation to LEN parameters, which can be any _n_ >= 0 in count for a derived type. (FIR accepts a dynamic number of LEN parameters on various ops because of this dependent type issue.)